src/overlays/actors/ovl_En_Heishi2/z_en_heishi2.c

```diff
@@ -21,77 +67,880 @@ const ActorInit En_Heishi2_InitVars = {
+            angleDiff = this->actor.rotTowardsLinkY - this->actor.shape.rot.y;
+            if (angleDiff >= 0) {
+                newAngleDiff = angleDiff;
+            } else {
+                newAngleDiff = -angleDiff;
+            }
+            if (!(120.0f < this->actor.xzDistanceFromLink)) {
+                if (newAngleDiff < 0x4300) {
+                    func_8002F2F4(thisx, globalCtx);
+                }
+            }
+        }
+    }
+}
+#else
 #pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/func_80A5475C.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/func_80A54954.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/func_80A549E8.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/EnHeishi2_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/func_80A54BC0.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/func_80A54C34.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/func_80A54C6C.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi2/EnHeishi2_Draw.s")
+#endif
+
+void func_80A54954(EnHeishi2* this, GlobalContext* globalCtx) {
+    f32 frameCount = SkelAnime_GetFrameCount(&D_06005C30.genericHeader);
+    SkelAnime_ChangeAnim(&this->skelAnime, &D_06005C30, 1.0f, 0.0f, (s16)frameCount, 0, -10.0f);
+    this->actionFunc = func_80A549E8;
+}
+
+void func_80A549E8(EnHeishi2* this, GlobalContext* globalCtx) {
+    SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+    if (this->unk_300 == func_8010BDBC(&globalCtx->msgCtx)) {
+        if (func_80106BC8(globalCtx) != 0) {
+            func_80106CCC(globalCtx);
+            if (this->initParams == 2) {
+                this->actionFunc = func_80A531E4;
+            }
+            if (this->initParams == 5) {
+                this->actionFunc = func_80A53908;
+            }
+        }
+    }
+}
+
+void EnHeishi2_Update(EnHeishi2* this, GlobalContext* globalCtx) {
+    ColliderCylinder* collider;
+    EnHeishi2* heishi2Temp;
+    Actor* thisx = &this->actor;
+    s32 phi_v1;
+
+    Actor_SetHeight(thisx, this->unk_2E0);
+    if ((this->initParams == 2) || (this->initParams == 5)) {
+        this->actor.posRot2.pos.y = 70.0f;
+        Actor_SetHeight(thisx, 70.0f);
+        func_80038290(globalCtx, thisx, &this->unk_260, &this->unk_26C, thisx->posRot2.pos);
+    }
+
+    this->unk_2FC++;
+    phi_v1 = 0;
+    heishi2Temp = this;
+
+    do {
+        phi_v1 += 2;
+        if (heishi2Temp->gateTimer != 0) {
+            heishi2Temp->gateTimer--;
+        }
+        heishi2Temp = (EnHeishi2*)&(heishi2Temp->actor.type); // what is going on here ???
+    } while (phi_v1 != 10);
+    this->actionFunc(this, globalCtx);
+    Actor_MoveForward(thisx);
+    if (this->initParams != 6) {
+        if (this->initParams != 9) {
+            func_8002E4B4(globalCtx, thisx, 10.0f, 10.0f, 30.0f, 0x1D);
```

Comment:
this looks like its probably a switch where case 6 and 9 break and defaults to this